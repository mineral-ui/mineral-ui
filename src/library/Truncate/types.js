/* @flow */

export type TruncateProps = {
  children?: React$Node
};

export type TruncateState = {
  showTooltip: boolean
};
