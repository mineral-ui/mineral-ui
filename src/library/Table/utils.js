/* @flow */
import type { Rows, TableProps } from './types';

export const getColumns = ({ columns, data }: TableProps) => {
  return (
    columns ||
    (data[0]
      ? Object.keys(data[0]).reduce((acc, cell) => {
          acc.push({ content: cell, key: cell });
          return acc;
        }, [])
      : [])
  );
};

export const getComparators = ({ columns }: TableProps) => {
  const comparators =
    columns &&
    columns.reduce((acc, column) => {
      const { key, sortComparator } = column;
      if (sortComparator) {
        acc[key] = sortComparator;
      }
      return acc;
    }, {});

  return comparators && Object.keys(comparators).length
    ? comparators
    : undefined;
};

export const getSelectableRows = (rows: Rows) =>
  rows.filter((row) => !row.disabled);

export const getSortable = ({
  columns,
  defaultSort,
  sort,
  sortable
}: TableProps) =>
  Boolean(
    defaultSort ||
      sort ||
      sortable ||
      (columns && columns.some((column) => column.sortable))
  );
