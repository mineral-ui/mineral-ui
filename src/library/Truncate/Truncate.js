/* @flow */
import React, { PureComponent } from 'react';
import debounce from 'lodash.debounce';
import { ellipsis } from 'polished';
import { createStyledComponent } from '../styles';
import _Tooltip from '../Tooltip';

type Props = {
  /** Content of Truncate */
  children?: React$Node
};

type State = {
  showTooltip: boolean
};

const styles = {
  root: ({ theme: baseTheme }) => ({
    pointerEvents: 'all', // Necessary because of Button Inner's pointerEvents: none
    ...ellipsis('100%'),

    '&:focus': {
      outline: `1px solid ${baseTheme.color_theme}`,
      outlineOffset: -1
    }
  }),
  tooltip: {
    // This style is only necessary because we're using Truncate as a Button child
    whiteSpace: 'normal'
  }
};

const Root = createStyledComponent('span', styles.root, {
  displayName: 'Truncate'
});
const Tooltip = createStyledComponent(_Tooltip, styles.tooltip);

export default class Truncate extends PureComponent<Props, State> {
  state = {
    showTooltip: false
  };

  root: ?HTMLElement;

  setRootRef = (node: HTMLElement) => {
    this.root = node;
  };

  toggleTooltip = () => {
    const rootNode = this.root;

    if (rootNode) {
      const offsetWidth = rootNode.offsetWidth + 1; // `+ 1` necessary for Edge & IE11
      const showTooltip = Boolean(rootNode.scrollWidth > offsetWidth);
      if (this.state.showTooltip !== showTooltip) {
        this.setState({
          showTooltip
        });
      }
    }
  };

  componentDidMount() {
    this.toggleTooltip();
  }

  componentDidUpdate() {
    this.toggleTooltip();
  }
  handleWindowResize = debounce(this.toggleTooltip, 100);

  render() {
    const { children, ...restProps } = this.props;
    const rootProps = {
      innerRef: this.setRootRef,
      ...restProps
    };

    const content = <Root {...rootProps}>{children}</Root>;
    return this.root && this.state.showTooltip ? (
      <Tooltip content={this.root.textContent}>{content}</Tooltip>
    ) : (
      content
    );
  }
}
