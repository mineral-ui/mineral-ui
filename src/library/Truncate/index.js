/* @flow */
export { default } from './Truncate';
