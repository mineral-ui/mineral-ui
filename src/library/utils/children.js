/* @flow */
import { Children } from 'react';

export function findByType(children: React$Node, type: React$ComponentType<*>) {
  let match;

  Children.forEach(children, (child) => {
    if (!match && child && child.type === type) {
      match = child;
    }
  });

  return match;
}

export function findAllByType(
  children: React$Node,
  type: React$ComponentType<*>
) {
  return Children.map(children, (child) => {
    if (child && child.type === type) {
      return child;
    }
  });
}
