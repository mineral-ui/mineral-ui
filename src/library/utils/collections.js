/* @flow */

export const fromArray = <T>(array: Array<T>): Set<T> =>
  array.reduce((acc, value) => {
    acc.add(value);
    return acc;
  }, new Set());

// Recursively converts an object of arrays into an object of sets.
// Sets are faster for containment checks.
export const settify = (input: any) =>
  Object.keys(input).reduce((acc, key) => {
    const value = input[key];
    acc[key] = Array.isArray(value) ? fromArray(value) : settify(value);
    return acc;
  }, {});

export const toArray = (value: any) => (Array.isArray(value) ? value : [value]);
