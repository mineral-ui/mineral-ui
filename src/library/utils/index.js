/* @flow */
export * from './children';
export { default as composeEventHandlers } from './composeEventHandlers';
export * from './generateId';
export { default as isRenderProp } from './isRenderProp';
export { default as isValidProp } from './isValidProp';
export { isDevelopment, isProduction } from './nodeEnv';
export { default as reactProps } from './reactProps';
export { fromArray, settify, toArray } from './collections';
