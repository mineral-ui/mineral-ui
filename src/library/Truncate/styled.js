/* @flow */
import styled from '@emotion/styled';
import { ellipsis } from 'polished';
import _Tooltip from '../Tooltip';
import { ie10Plus } from '../utils/cssSelectors';

import type { StyledComponent } from '@emotion/styled-base/src/utils';

export const TruncateRoot: StyledComponent<{ [key: string]: any }> = styled(
  'span'
)(({ showTooltip, theme: baseTheme }) => ({
  ...(showTooltip ? { pointerEvents: 'all' } : undefined), // Necessary because of Button Inner's pointerEvents: none
  ...ellipsis('100%'),

  '&:focus': {
    outline: `1px solid ${baseTheme.borderColor_theme_focus}`,
    outlineOffset: -1,

    [ie10Plus]: {
      outline: 'none'
    }
  }
}));

export const Tooltip: StyledComponent<{ [key: string]: any }> = styled(
  _Tooltip
)({
  // This style is only necessary because we're using Truncate as a Button child
  whiteSpace: 'normal'
});
