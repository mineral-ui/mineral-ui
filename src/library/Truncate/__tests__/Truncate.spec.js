/* @flow */
import React from 'react';
import { shallow } from 'enzyme';
import Truncate from '../Truncate';
import { mountInThemeProvider } from '../../../../utils/enzymeUtils';

/*
 * [1] This getter does not exist on the HTMLElement.prototype in JSDOM, so we
 *     must mock it on the global. `offsetWidth`, also used by Truncate, always
 *     returns 0, which is conveniently useful.
 * [2] Then we must clean up our global mocking in [1]
 */

const defaultProps = {
  children: 'Long text string that hopefully truncates'
};

function shallowTruncate(props = {}) {
  const TruncateProps = {
    ...defaultProps,
    ...props
  };
  return shallow(<Truncate {...TruncateProps} />);
}

describe('Truncate', () => {
  it('renders', () => {
    const Truncate = shallowTruncate();

    expect(Truncate.exists()).toEqual(true);
  });

  it('does not render tooltip when not truncated', () => {
    const [, truncate] = mountInThemeProvider(<Truncate {...defaultProps} />);

    expect(truncate).toMatchSnapshot();
  });

  it('renders a tooltip when truncated', () => {
    // [1]
    // $FlowFixMe
    Object.defineProperties(window.HTMLElement.prototype, {
      scrollWidth: {
        // Note that this mocked getter is designed to always trigger truncation
        // because it's greater than offsetWidth
        get: () => 10,
        configurable: true
      }
    });

    const [, truncate] = mountInThemeProvider(<Truncate {...defaultProps} />);

    // [2]
    // $FlowFixMe
    Object.defineProperties(window.HTMLElement.prototype, {
      scrollWidth: {
        get: undefined
      }
    });

    expect(truncate).toMatchSnapshot();
  });
});
