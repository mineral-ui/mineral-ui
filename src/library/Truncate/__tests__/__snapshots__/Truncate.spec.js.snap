// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Truncate does not render tooltip when not truncated 1`] = `
.emotion-0 {
  pointer-events: all;
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}

.emotion-0:focus {
  outline: 1px solid #3272d9;
  outline-offset: -1px;
}

<Truncate>
  <Truncate
    innerRef={[Function]}
  >
    <span
      className="emotion-0"
    >
      Long text string that hopefully truncates
    </span>
  </Truncate>
</Truncate>
`;

exports[`Truncate renders a tooltip when truncated 1`] = `
.emotion-0 {
  pointer-events: all;
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}

.emotion-0:focus {
  outline: 1px solid #3272d9;
  outline-offset: -1px;
}

.emotion-5 {
  white-space: normal;
}

.emotion-3 {
  display: inline-block;
  white-space: normal;
}

.emotion-1 {
  display: inline-block;
}

<Truncate>
  <Styled(Tooltip)
    content="Long text string that hopefully truncates"
    hasArrow={true}
  >
    <Tooltip
      className="emotion-5"
      content="Long text string that hopefully truncates"
      hasArrow={true}
    >
      <WithTheme(Themed(Popover))
        className="emotion-5"
        content={[Function]}
        focusTriggerOnClose={false}
        hasArrow={true}
        id="tooltip-1"
        isOpen={false}
        onClose={[Function]}
        onOpen={[Function]}
      >
        <Themed(Popover)
          className="emotion-5"
          content={[Function]}
          focusTriggerOnClose={false}
          hasArrow={true}
          id="tooltip-1"
          isOpen={false}
          onClose={[Function]}
          onOpen={[Function]}
        >
          <ThemeProvider>
            <ThemeProvider>
              <Popover
                className="emotion-5"
                content={[Function]}
                focusTriggerOnClose={false}
                hasArrow={true}
                id="tooltip-1"
                isOpen={false}
                onClose={[Function]}
                onOpen={[Function]}
                placement="bottom"
              >
                <Popover
                  className="emotion-5"
                  content={[Function]}
                  focusTriggerOnClose={false}
                  hasArrow={true}
                  id="tooltip-1"
                  isOpen={false}
                  onClose={[Function]}
                  onOpen={[Function]}
                  placement="bottom"
                  tag="span"
                >
                  <Popover
                    className="emotion-3"
                    id="tooltip-1"
                    tag="span"
                  >
                    <span
                      className="emotion-3"
                      id="tooltip-1"
                    >
                      <PopoverTrigger
                        aria-describedby="tooltip-1-content"
                        aria-owns="tooltip-1-content"
                        innerRef={[Function]}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        role="button"
                        tabIndex={0}
                      >
                        <PopoverTrigger
                          component="span"
                        >
                          <PopoverTrigger
                            className="emotion-1"
                            component="span"
                          >
                            <span
                              className="emotion-1"
                            >
                              <Truncate
                                aria-describedby="tooltip-1-content"
                                aria-owns="tooltip-1-content"
                                innerRef={[Function]}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onFocus={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                role="button"
                                tabIndex={0}
                              >
                                <span
                                  aria-describedby="tooltip-1-content"
                                  aria-owns="tooltip-1-content"
                                  className="emotion-0"
                                  onBlur={[Function]}
                                  onClick={[Function]}
                                  onFocus={[Function]}
                                  onMouseEnter={[Function]}
                                  onMouseLeave={[Function]}
                                  role="button"
                                  tabIndex={0}
                                >
                                  Long text string that hopefully truncates
                                </span>
                              </Truncate>
                            </span>
                          </PopoverTrigger>
                        </PopoverTrigger>
                      </PopoverTrigger>
                    </span>
                  </Popover>
                </Popover>
              </Popover>
            </ThemeProvider>
          </ThemeProvider>
        </Themed(Popover)>
      </WithTheme(Themed(Popover))>
    </Tooltip>
  </Styled(Tooltip)>
</Truncate>
`;
